### PR TITLE
Add GC finalization hook

### PR DESCRIPTION
**Pull request by dcodeIO**
Related: https://github.com/AssemblyScript/assemblyscript/issues/383

This is a quick experiment for a finalization hook when managed objects become collected. A lot can go wrong here, especially when finalization messes with MM/GC state before freeing, so this is as low-level and unsafe as it gets. Doesn't bother with registering destructors (needs compiler support or a virtual destructor table, and suffers from the same challenges) but might be a starting point for anyone interested in playing with finalization in general / inform future design. Wondering if a mechanism along the lines of JS weakref can be implemented on top.

### CODE REVIEW

File: tests/compiler/rt/finalize.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    var map = this.map;
+    if (map.has(ptr)) {
+      let value = map.get(ptr);
+      this.fn(value);
+      map.delete(ptr);
+    }
+  }
+}
+
+var registry = new FinalizationRegistry(value => {
```
Comment (jtenner) [inline tests/compiler/rt/finalize.ts:23]: So this is how we should hook into object finalization? Any tips or important things we should be aware of? Also, where should this go in the documentation?

Comment (dcodeIO) [reply tests/compiler/rt/finalize.ts:23]: That's really just an initial idea along the lines of [JS weakrefs](https://github.com/tc39/proposal-weakrefs). The advantage this has over destructors is that the finalization function isn't called with the actual object being free'd, which would be super unsafe, but with the registered value. Nothing set in stone yet.

**Comment by dcodeIO** [comment]
Keeping this PR open as part of [2020 vacuum](https://github.com/AssemblyScript/assemblyscript/issues/1277) as it hasn't yet received enough feedback to make a proper decision just yet. Also considering, in case it appears useful, to merge just the `__finalize` runtime hook as a low-level building block to encourage experimentation. It's opt-in anyway and doesn't affect code not using it.

**Comment by dcodeIO** [comment]
Thinking about it a little more, since it doesn't affect existing code and is opt-in, let's just merge this. Shouldn't be a burden due to its simplicity, and enables everyone to experiment with finalization to aid future design of a proper API. Updated the test to check just `__finalize` without any design assumptions.

**Comment by jtenner** [comment]
How do we use this officially? Is there anything in the docs about it?

**Comment by dcodeIO** [comment]
An example of how to use it is [the test file](https://github.com/AssemblyScript/assemblyscript/blob/master/tests/compiler/rt/finalize.ts). Essentially, one has to add

```ts
@global function __finalize(ptr: usize): void {
  ...
}
```

as an additional runtime hook, which is then called with the pointer to the object being free'd right before it is. Not the final API of course (that'd be something more along the lines of `FinalizationRegistry`), just the integration point. Not sure if it deserves to be included in the docs just yet, but perhaps we can mention it [within RT internals](https://github.com/AssemblyScript/assemblyscript/tree/master/std/assembly/rt).

**Comment by jtenner** [comment]
This is amazing. Thank you.

**Pull request merged**